power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -578,6 +583,7 @@ class TapPositionOptimizerImpl<std::tuple<TransformerTypes...>, StateCalculator,
     using TransformerRanker = TransformerRanker_;
 
   private:
+    mutable std::vector<uint64_t> max_tap_ranges_per_rank{};
```

Comment:
tap_min and tap_max are not going to change for a input_data.
I wonder if we can just have tap_range at `RegulatedObjects` instead. They can be calculated in `retrieve_regulator_info()`.
And then the max iteration of each rank in `RankedTransformerGroups`. They can be calculated at `rank_transformers()`

Reply:
> tap_min and tap_max are not going to change for a input_data.
I wonder if we can just have tap_range at RegulatedObjects instead. They can be calculated in retrieve_regulator_info().

`tap_range` is already available in transformers as a function call. In terms of number of calls, they are the same whether they live as a member or as a function call, because only once will they be required throughout the code (here). If we put everything inside `RegulatedObjects`, which is container of indices, I am not sure if this is making things less scattered or more so XD

> And then the max iteration of each rank in RankedTransformerGroups. They can be calculated at rank_transformers()

Having this per rank information inside `RankedTransformerGroups` indeed makes sense, and that was my original code as well. After bumping into some template mismatch, I figure it might also make sense that this is a per optimizer requirement, and not necessarily order group property (which is again groups of indices).